client/beefy/src/round.rs

```diff
@@ -111,60 +115,77 @@ where
 			debug!(target: LOG_TARGET, "🥩 received vote for old stale round {:?}, ignoring", num);
-			false
-		} else if !self.validators().iter().any(|id| vote.0 == *id) {
+			return VoteImportResult::Stale
+		} else if vote.commitment.validator_set_id != self.validator_set_id() {
+			debug!(
+				target: LOG_TARGET,
+				"🥩 expected set_id {:?}, ignoring vote {:?}.",
+				self.validator_set_id(),
+				vote,
+			);
+			return VoteImportResult::Invalid
+		} else if !self.validators().iter().any(|id| &vote.id == id) {
 			debug!(
 				target: LOG_TARGET,
 				"🥩 received vote {:?} from validator that is not in the validator set, ignoring",
 				vote
 			);
-			false
+			return VoteImportResult::Invalid
+		}
+
+		if let Some(previous_vote) = self.previous_votes.get(&equivocation_key) {
+			// is the same public key voting for a different payload?
+			if previous_vote.commitment.payload != vote.commitment.payload {
+				debug!(
+					target: LOG_TARGET,
+					"🥩 detected equivocated vote: 1st: {:?}, 2nd: {:?}", previous_vote, vote
+				);
+				// TODO: build `EquivocationProof` and return it here.
+				return VoteImportResult::Equivocation
+			}
 		} else {
-			self.rounds.entry(round.clone()).or_default().add_vote(vote, self_vote)
+			// this is the first vote sent by `id` for `num`, all good
+			self.previous_votes.insert(equivocation_key, vote.clone());
 		}
+
+		// add valid vote
+		let round = self.rounds.entry(vote.commitment.clone()).or_default();
+		if round.add_vote((vote.id, vote.signature)) &&
+			round.is_done(threshold(self.validator_set.len()))
+		{
+			if let Some(round) = self.rounds.remove_entry(&vote.commitment) {
+				return VoteImportResult::RoundConcluded(self.signed_commitment(round))
+			}
+		}
+		VoteImportResult::Ok
 	}
 
-	pub(crate) fn should_conclude(
+	fn signed_commitment(
 		&mut self,
-		round: &(P, NumberFor<B>),
-	) -> Option<Vec<Option<Signature>>> {
-		let done = self
-			.rounds
-			.get(round)
-			.map(|tracker| tracker.is_done(threshold(self.validator_set.len())))
-			.unwrap_or(false);
-		trace!(target: LOG_TARGET, "🥩 Round #{} done: {}", round.1, done);
-
-		if done {
-			let signatures = self.rounds.remove(round)?.votes;
-			Some(
-				self.validators()
-					.iter()
-					.map(|authority_id| signatures.get(authority_id).cloned())
-					.collect(),
-			)
-		} else {
-			None
-		}
+		round: (Commitment<NumberFor<B>>, RoundTracker),
+	) -> SignedCommitment<NumberFor<B>, Signature> {
+		let votes = round.1.votes;
+		let signatures = self
+			.validators()
+			.iter()
+			.map(|authority_id| votes.get(authority_id).cloned())
+			.collect();
+		SignedCommitment { commitment: round.0, signatures }
 	}
 
 	pub(crate) fn conclude(&mut self, round_num: NumberFor<B>) {
 		// Remove this and older (now stale) rounds.
-		self.rounds.retain(|&(_, number), _| number > round_num);
+		self.rounds.retain(|commitment, _| commitment.block_number > round_num);
+		self.previous_votes.retain(|&(_, number), _| number > round_num);
```

Comment:
Not sure if in the future we might want to keep these around for longer. But let's keep it like this for now.